### DISCUSSION TITLE
Language INI files are extremely costly

### DISCUSSION THREAD

**Discussion by dgrammatiko** [joomla/joomla-cms#38589] [2. Cold Features]
The language strings come as INI files. There's a cost of reading/parsing/converting to array for each of these files.
This can be done ONCE, after install/update/remove and save the file as a php returning an array. In short at runtime the app will have to only read a file and merge the array returned from the language file to the array of translation strings. The INI could stay as fallback. The load sequence also could be simplified. Bonus since the files will be .php they could benefit further from OPCache...

Why?

The current architecture is very inefficient both on terms of wasted cpu cycles and excessive energy consumption.

If you want to go one step further you could have per application (site/administrator) a file that has all the sys.ini files concatenated. It increases the memory but reduces I/O operations

Again this is touching architecture so something for @nibra @HLeithner @laoneo

**Comment by Fedik** [comment]
>There's a cost of reading/parsing/converting to array for each of these files. This can be done ONCE

That sounds valid. Would be interesting to see numbers, that can give motivation to someone to code it.

> The current architecture is very inefficient 

Please do not run to the edges.

**Reply by dgrammatiko** [reply]
> Please do not run to the edges.

I think you misinterpreted what I wrote. In J4 there is this `administrator/cache/autoload_psr4.php` which is generated after the installation or each update (core or extension). The job (reading/parsing/creating the array) is done once. This is efficient and it's the opposite of what is happening right now with the language files (thus the very inefficient)
